src/de.rs

```diff
@@ -0,0 +1,491 @@
+use pyo3::type_object::PyTypeInfo;
+use pyo3::types::*;
+use pyo3::{PyObject, Python};
+use serde::{de, Deserialize};
+
+use crate::error::{PythonizeError, Result};
+
+pub fn depythonize<'de, T>(py: Python, obj: PyObject) -> Result<T>
+where
+    T: Deserialize<'de>,
+{
+    let mut depythonizer = Depythonizer::from_object(py, obj);
+    T::deserialize(&mut depythonizer)
+}
+
+#[derive(Debug)]
+enum GetItemKey {
+    Key(PyObject),
+    Index(isize),
+    None,
+}
+
+pub struct Depythonizer<'py> {
+    input: PyObject,
+    current: GetItemKey,
+    py: Python<'py>,
+}
+
+impl<'py> Depythonizer<'py> {
+    pub fn from_object(py: Python<'py>, input: PyObject) -> Self {
+        Depythonizer {
+            input,
+            current: GetItemKey::None,
+            py,
+        }
+    }
+
+    fn get_item(&self) -> Result<Option<PyObject>> {
```

Comment:
Judging by uses of this, I think you might find it more ergonomic to make a variant of this:

```suggestion
    fn get_item_or_missing(&self) -> Result<&PyAny> {
```

and if the dict/list lookup fails, return `Err(PythonizeError::missing(&self.current))` directly from this function.

Also, returning `&PyAny` here instead of `PyObject` can potentially avoid increasing reference counts.